File: frontend/src/types/datasource.ts
/**
 * データソース関連の型定義
 */

/** データベースタイプ */
export type DBType = "postgresql" | "mysql" | "oracle" | "sqlserver";

/** データベースタイプの表示名 */
export const DB_TYPE_LABELS: Record<DBType, string> = {
  postgresql: "PostgreSQL",
  mysql: "MySQL",
  oracle: "Oracle",
  sqlserver: "SQL Server",
};

/** デフォルトポート番号 */
export const DEFAULT_PORTS: Record<DBType, number> = {
  postgresql: 5432,
  mysql: 3306,
  oracle: 1521,
  sqlserver: 1433,
};

/** データソース */
export interface DataSource {
  id: number;
  name: string;
  db_type: DBType;
  host: string;
  port: number;
  database_name: string;
  username: string;
  created_by: number;
  created_at: string;
  updated_at: string;
}

/** データソース作成リクエスト */
export interface CreateDataSourceRequest {
  name: string;
  db_type: DBType;
  host: string;
  port: number;
  database_name: string;
  username: string;
  password: string;
}

/** データソース更新リクエスト */
export interface UpdateDataSourceRequest {
  name?: string;
  host?: string;
  port?: number;
  database_name?: string;
  username?: string;
  password?: string;
}

/** テスト接続リクエスト */
export interface TestConnectionRequest {
  db_type: DBType;
  host: string;
  port: number;
  database_name: string;
  username: string;
  password: string;
}

/** テスト接続レスポンス */
export interface TestConnectionResponse {
  success: boolean;
  message: string;
}

/** データソース一覧レスポンス */
export interface DataSourceListResponse {
  data_sources: DataSource[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    total_pages: number;
  };
}

/** テーブル情報 */
export interface TableInfo {
  name: string;
  schema?: string;
}

/** テーブル一覧レスポンス */
export interface TableListResponse {
  tables: TableInfo[];
}

/** カラム情報 */
export interface ColumnInfo {
  name: string;
  data_type: string;
  is_nullable: boolean;
  is_primary_key: boolean;
  default_value?: string;
}

/** カラム一覧レスポンス */
export interface ColumnListResponse {
  columns: ColumnInfo[];
}

/** 外部アプリ作成リクエストのフィールド */
export interface CreateExternalFieldRequest {
  source_column_name: string;
  field_code: string;
  field_name: string;
  field_type: string;
  options?: Record<string, unknown>;
  required: boolean;
  display_order: number;
}

/** 外部アプリ作成リクエスト */
export interface CreateExternalAppRequest {
  name: string;
  description?: string;
  icon?: string;
  data_source_id: number;
  source_table_name: string;
  fields: CreateExternalFieldRequest[];
}

/** データ型マッピング（外部DBカラム型 → アプリフィールド型） */
export function mapDataTypeToFieldType(dataType: string): string {
  const normalizedType = dataType.toLowerCase();

  // 数値型
  if (
    /^(int|bigint|smallint|tinyint|decimal|numeric|float|double|real|number)/.test(
      normalizedType
    )
  ) {
    return "number";
  }

  // 日付時刻型
  if (/^(datetime|timestamp)/.test(normalizedType)) {
    return "datetime";
  }
  if (/^date$/.test(normalizedType)) {
    return "date";
  }

  // テキスト型（長い）
  if (/^(text|clob|longtext|mediumtext|ntext)/.test(normalizedType)) {
    return "textarea";
  }

  // ブール型
  if (/^(bool|boolean|bit)/.test(normalizedType)) {
    return "checkbox";
  }

  // JSON型
  if (/^json/.test(normalizedType)) {
    return "textarea";
  }

  // デフォルトはテキスト
  return "text";
}

File: backend/internal/models/datasource.go
package models

import (
	"time"

	"github.com/uptrace/bun"
)

// DBType データベースタイプの定数
type DBType string

const (
	DBTypePostgreSQL DBType = "postgresql"
	DBTypeMySQL      DBType = "mysql"
	DBTypeOracle     DBType = "oracle"
	DBTypeSQLServer  DBType = "sqlserver"
)

// ValidDBTypes 有効なデータベースタイプ一覧
var ValidDBTypes = []DBType{
	DBTypePostgreSQL,
	DBTypeMySQL,
	DBTypeOracle,
	DBTypeSQLServer,
}

// IsValidDBType データベースタイプが有効かどうかを検証する
func IsValidDBType(dbType string) bool {
	for _, t := range ValidDBTypes {
		if string(t) == dbType {
			return true
		}
	}
	return false
}

// DataSource 外部データベース接続情報を表す構造体
type DataSource struct {
	bun.BaseModel `bun:"table:data_sources,alias:ds"`

	ID                uint64    `bun:"id,pk,autoincrement" json:"id"`
	Name              string    `bun:"name,notnull,unique" json:"name"`
	DBType            DBType    `bun:"db_type,notnull" json:"db_type"`
	Host              string    `bun:"host,notnull" json:"host"`
	Port              int       `bun:"port,notnull" json:"port"`
	DatabaseName      string    `bun:"database_name,notnull" json:"database_name"`
	Username          string    `bun:"username,notnull" json:"username"`
	EncryptedPassword string    `bun:"encrypted_password,notnull" json:"-"` // JSONには出力しない
	CreatedBy         uint64    `bun:"created_by,notnull" json:"created_by"`
	CreatedAt         time.Time `bun:"created_at,notnull,default:current_timestamp" json:"created_at"`
	UpdatedAt         time.Time `bun:"updated_at,notnull,default:current_timestamp" json:"updated_at"`
	Creator           *User     `bun:"rel:belongs-to,join:created_by=id" json:"creator,omitempty"`
}

// CreateDataSourceRequest データソース作成リクエストの構造体
type CreateDataSourceRequest struct {
	Name         string `json:"name" validate:"required,min=1,max=100"`
	DBType       string `json:"db_type" validate:"required,oneof=postgresql mysql oracle sqlserver"`
	Host         string `json:"host" validate:"required,min=1,max=255"`
	Port         int    `json:"port" validate:"required,min=1,max=65535"`
	DatabaseName string `json:"database_name" validate:"required,min=1,max=100"`
	Username     string `json:"username" validate:"required,min=1,max=100"`
	Password     string `json:"password" validate:"required,min=1"`
}

// UpdateDataSourceRequest データソース更新リクエストの構造体
type UpdateDataSourceRequest struct {
	Name         string `json:"name" validate:"omitempty,min=1,max=100"`
	Host         string `json:"host" validate:"omitempty,min=1,max=255"`
	Port         int    `json:"port" validate:"omitempty,min=1,max=65535"`
	DatabaseName string `json:"database_name" validate:"omitempty,min=1,max=100"`
	Username     string `json:"username" validate:"omitempty,min=1,max=100"`
	Password     string `json:"password" validate:"omitempty,min=1"` // 空の場合は更新しない
}

// TestConnectionRequest テスト接続リクエストの構造体
type TestConnectionRequest struct {
	DBType       string `json:"db_type" validate:"required,oneof=postgresql mysql oracle sqlserver"`
	Host         string `json:"host" validate:"required,min=1,max=255"`
	Port         int    `json:"port" validate:"required,min=1,max=65535"`
	DatabaseName string `json:"database_name" validate:"required,min=1,max=100"`
	Username     string `json:"username" validate:"required,min=1,max=100"`
	Password     string `json:"password" validate:"required,min=1"`
}

// DataSourceResponse データソースのレスポンス構造体（パスワードは含まない）
type DataSourceResponse struct {
	ID           uint64    `json:"id"`
	Name         string    `json:"name"`
	DBType       DBType    `json:"db_type"`
	Host         string    `json:"host"`
	Port         int       `json:"port"`
	DatabaseName string    `json:"database_name"`
	Username     string    `json:"username"`
	CreatedBy    uint64    `json:"created_by"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
}

// ToResponse DataSourceをDataSourceResponseに変換する
func (ds *DataSource) ToResponse() *DataSourceResponse {
	return &DataSourceResponse{
		ID:           ds.ID,
		Name:         ds.Name,
		DBType:       ds.DBType,
		Host:         ds.Host,
		Port:         ds.Port,
		DatabaseName: ds.DatabaseName,
		Username:     ds.Username,
		CreatedBy:    ds.CreatedBy,
		CreatedAt:    ds.CreatedAt,
		UpdatedAt:    ds.UpdatedAt,
	}
}

// DataSourceListResponse データソース一覧のレスポンス構造体
type DataSourceListResponse struct {
	DataSources []DataSourceResponse `json:"data_sources"`
	Pagination  *Pagination          `json:"pagination"`
}

// TableType テーブルタイプの定数
type TableType string

const (
	TableTypeTable TableType = "TABLE"
	TableTypeView  TableType = "VIEW"
)

// TableInfo テーブル情報の構造体
type TableInfo struct {
	Name   string    `json:"name"`
	Schema string    `json:"schema,omitempty"`
	Type   TableType `json:"type"`
}

// TableListResponse テーブル一覧のレスポンス構造体
type TableListResponse struct {
	Tables []TableInfo `json:"tables"`
}

// ColumnInfo カラム情報の構造体
type ColumnInfo struct {
	Name         string `json:"name"`
	DataType     string `json:"data_type"`
	IsNullable   bool   `json:"is_nullable"`
	IsPrimaryKey bool   `json:"is_primary_key"`
	DefaultValue string `json:"default_value,omitempty"`
}

// ColumnListResponse カラム一覧のレスポンス構造体
type ColumnListResponse struct {
	Columns []ColumnInfo `json:"columns"`
}

// TestConnectionResponse テスト接続のレスポンス構造体
type TestConnectionResponse struct {
	Success bool   `json:"success"`
	Message string `json:"message"`
}

// GetDefaultPort データベースタイプに応じたデフォルトポートを返す
func GetDefaultPort(dbType DBType) int {
	switch dbType {
	case DBTypePostgreSQL:
		return 5432
	case DBTypeMySQL:
		return 3306
	case DBTypeOracle:
		return 1521
	case DBTypeSQLServer:
		return 1433
	default:
		return 0
	}
}
